File: two-pointers/844/844.py
# 1
class Solution:
    def backspaceCompare(self, S: str, T: str) -> bool:
        def build(string):
            result = []
            for ch in string:
                if ch != "#":
                    result.append(ch)
                else:
                    if result:
                        result.pop()

            return "".join(result)

        return build(S) == build(T)


# 2
class Solution:
    def backspaceCompare(self, S: str, T: str) -> bool:
        def findNonSkippingPoint(string, i):
            skip = 0
            while i >= 0:
                if string[i] == "#":
                    skip += 1
                    i -= 1
                elif skip > 0:
                    skip -= 1
                    i -= 1
                else:
                    break
            return i

        s = len(S) - 1
        t = len(T) - 1

        while s >= 0 or t >= 0:
            s = findNonSkippingPoint(S, s)
            t = findNonSkippingPoint(T, t)

            if s >= 0 and t >= 0 and S[s] != T[t]:
                return False
            if (s >= 0) != (t >= 0):
                return False
            s -= 1
            t -= 1
        return True

File: two-pointers/844/844.js
//#1
var backspaceCompare = function(S, T) {
  return help(S) === help(T);
};

const help = string => {
  let result = [];
  for (ch of string) {
    if (ch === "#") {
      result.pop();
    } else {
      result.push(ch);
    }
  }
  return result.join();
};

//#2
var backspaceCompare = function(S, T) {
  let S_pointer = S.length - 1;
  let T_pointer = T.length - 1;
  while (S_pointer >= 0 || T_pointer >= 0) {
    S_pointer = findNonSkippingPoint(S, S_pointer);
    T_pointer = findNonSkippingPoint(T, T_pointer);
    //if there's still "meat" left after all the skipping, check if  the remaining characters are the same starting from the back.If not, return false
    if (S_pointer >= 0 && T_pointer >= 0 && S[S_pointer] !== T[T_pointer]) {
      return false;
    }
    //if one of the string is exhausted before the other, return false. Can't just simply use S_pointer<0||T_pointer<0, because there are situations when both strings are exhausted, e.g.("ab##","c#d#")
    if (S_pointer >= 0 !== T_pointer >= 0) {
      return false;
    }
    S_pointer--;
    T_pointer--;
  }
  return true;
};

//this helper method returns the "comparing" pointer after all the skipping
const findNonSkippingPoint = (string, i) => {
  let skip = 0;
  while (i >= 0) {
    if (string[i] == "#") {
      skip++;
      i--;
    } else if (skip > 0) {
      skip--;
      i--;
    } else {
      break;
    }
  }
  return i;
};
